Add filter bar to Choose Cloud Model page

jgravel opened the pull request:
SCRD-1712
* This pull request depends on the ardana-service change (https://gerrit.suse.provo.cloud/2153).

changes include:
- remove Help Choose and Select Template buttons
- show all templates instead of 5
- add 4 filter buttons and the filter function

also include fixes for lints complaints:
- Complete.js
- AssignServerRoles.js

jgravel committed:
Add filter bar to Choose Cloud Model page

The following are changes for SCRD-1712, which include:
- remove Help Choose and Select Template buttons
- show all templates instead of 5
- add 4 filter buttons and the filter function

also include fixes for lints complaints:
- Complete.js
- AssignServerRoles.js

GarySmith commented on src/pages/CloudModelPicker.js:
```diff
@@ -145,21 +130,137 @@ class CloudModelPicker extends BaseWizardPage {
+    // 'newFilters' keeps track of ALL filter options which changes every time the user makes a new
+    // selection by looking at metadata of all templates and sorting out the available values by
+    // filter type then displaying those values as filter options
+    let newFilters = JSON.parse(JSON.stringify(FILTERS));
```
This can't be right. It appears that this converting to a string and back again (YUCK) just to make a copy so that it can be modified in place. It would be more idiomatic to use FILTERS.map, which iterates through and returns a new array without touching the original

GarySmith commented on src/pages/CloudModelPicker.js:
```diff
@@ -20,30 +20,35 @@ import { fetchJson } from '../utils/RestUtils.js';
+import Dropdown from '../components/Dropdown.js';
+
+const FILTERS = [
+  {name: 'node-count', options: ['none', '30', '40']},
```
30 and 40 do not seem like correct node counts, but it's a little unclear what these represent. If they are maximum, then I believe that mid-size goes to 200, right? Will an entry-size support 30?
Also, if we are doing any kind of math on them, then numbers are more appropriate than string values

jgravel committed:
Update filter logic for node count and 'none' messages

jgravel commented:
Updated the code per Gary's comment (and the ardana-service code as well). Please take a look again.

jeremy-moffitt reviewed (changes requested):
When I set the first filter to 0-30 Compute nodes the following happens:
1. the midscale model I'd previously selected is still selected, just not visible
2. "Next" remains enabled, even though I've ostensibly selected a model thats not visible on the screen
3. the dropdown for compute node size no longer includes the >30 node option, instead offering only "0-30 Compute Nodes" and "Any Number of Compute Nodes".... if I select "Any Number of Compute Nodes", the choice for ">30 Compute Nodes" option returns

I think whats happening is that the filtering-out logic is excluding parallel choices... I think we'll need a different approach to this

jgravel committed:
Update display option logic per Jeremy's comments

jgravel commented:
Updated the logic to calculate filter options to display ALL options for the currently selected filter and to display only applicable options for the other three filters. Also updated the code to clear out the selected model if the selected model is filtered out and not showing on the screen.

jeremy-moffitt commented:
One minor nit while reviewing the other changes, might need some undefined protection, without the metadata change on the ardana-service I get "Cannot read property 'nodeCount' of undefined" ... not sure if we need to protect against that in case new models are added or not... could go either way

jeremy-moffitt commented:
per our conversation, theres a scenario where invalid options become available if the user interacts with the dropdowns in a certain order:
1. select 0-30 compute nodes
2. select KVM hypervisor
3. select VSA storage
4. in the hypervisor selector, click "any hypervisor"
5. open the hypervisor selector again, Ironic is now visible as an option, though it is not compatible with the combination of VSA and 0-30 compute nodes

jgravel committed:
Update filter logic to deal with inapplicable filter options

jgravel commented:
Updated filter logic to deal with inapplicable filter option and multiple filter selection, also add the ability to not show the filter bar when there's no metadata to create filter boxes (this is related to the problem that Jeremy reported about Cannot read property 'nodeCount' of undefined)

GarySmith commented:
Seems to work ok. Want to get this in before we move over to the new location.
